File: sdn-controller/test/unit/test_network_graph.ml
(* Unit Tests for Network Graph and Path Computation
   ==================================================
   These tests verify the core data structures and algorithms
   that the SDN controller relies on.

   We use Alcotest — a lightweight OCaml testing framework.
   (Later we'll add QCheck property-based tests in Phase 10.)

   TESTING PHILOSOPHY (for interview):
   - Unit tests check specific scenarios we can predict
   - Property tests (QCheck, Phase 10) check invariants that
     should hold for ALL inputs — much more powerful
   - Integration tests (later) check the full API end-to-end

   We test:
   1. Graph construction (adding devices and links)
   2. Path computation (Dijkstra correctness)
   3. Failover behavior (path after link failure)
   4. Traffic-class-specific routing (different weights)
   5. Edge cases (no path, single node, disconnected graph)
*)

open Topology.Network_graph
open Pathcomp.Path_engine

(* ============================================================
   Test Helpers
   ============================================================ *)

let default_metrics = {
  latency_us = 100.0;
  jitter_us = 10.0;
  utilization = 0.1;
  packet_loss = 0.0;
  last_updated = 0.0;
}

let make_device ~id ~role ~site ~loopback ~vendor =
  { id; name = id; role; site; loopback; vendor; asn = 65001 }

let make_link ~id ?(latency=100.0) ?(bw=10000) ?(cost=10) () = {
  id;
  src_interface = "eth0";
  dst_interface = "eth0";
  bandwidth_mbps = bw;
  state = Up;
  metrics = { default_metrics with latency_us = latency };
  ospf_cost = cost;
}

(* Build a simple 3-node linear topology: A -- B -- C *)
let build_linear_topology () =
  let g = empty () in
  let g = add_device g (make_device ~id:"A" ~role:CoreRouter
    ~site:"dc-east" ~loopback:"10.0.0.1" ~vendor:"cisco") in
  let g = add_device g (make_device ~id:"B" ~role:SpineSwitch
    ~site:"dc-east" ~loopback:"10.0.0.2" ~vendor:"arista") in
  let g = add_device g (make_device ~id:"C" ~role:EdgeRouter
    ~site:"colo-nyse" ~loopback:"10.0.0.3" ~vendor:"juniper") in
  let g = add_link g ~src:"A" ~dst:"B"
    (make_link ~id:"A--B" ~latency:50.0 ~cost:5 ()) in
  let g = add_link g ~src:"B" ~dst:"C"
    (make_link ~id:"B--C" ~latency:100.0 ~cost:10 ()) in
  g

(* Build a diamond topology: A --[fast]--> B ---> D
                             A --[slow]--> C ---> D
   This tests that Dijkstra picks the faster path *)
let build_diamond_topology () =
  let g = empty () in
  let g = add_device g (make_device ~id:"A" ~role:CoreRouter
    ~site:"dc-east" ~loopback:"10.0.0.1" ~vendor:"cisco") in
  let g = add_device g (make_device ~id:"B" ~role:CoreRouter
    ~site:"dc-east" ~loopback:"10.0.0.2" ~vendor:"cisco") in
  let g = add_device g (make_device ~id:"C" ~role:CoreRouter
    ~site:"dc-west" ~loopback:"10.0.0.3" ~vendor:"cisco") in
  let g = add_device g (make_device ~id:"D" ~role:EdgeRouter
    ~site:"colo-nyse" ~loopback:"10.0.0.4" ~vendor:"juniper") in
  (* Fast path: A -> B -> D (50 + 50 = 100μs) *)
  let g = add_link g ~src:"A" ~dst:"B"
    (make_link ~id:"A--B" ~latency:50.0 ~cost:5 ()) in
  let g = add_link g ~src:"B" ~dst:"D"
    (make_link ~id:"B--D" ~latency:50.0 ~cost:5 ()) in
  (* Slow path: A -> C -> D (200 + 200 = 400μs) *)
  let g = add_link g ~src:"A" ~dst:"C"
    (make_link ~id:"A--C" ~latency:200.0 ~cost:20 ()) in
  let g = add_link g ~src:"C" ~dst:"D"
    (make_link ~id:"C--D" ~latency:200.0 ~cost:20 ()) in
  g

(* ============================================================
   Graph Construction Tests
   ============================================================ *)

let test_empty_graph () =
  let g = empty () in
  Alcotest.(check int) "empty graph has no devices" 0
    (List.length (all_devices g));
  Alcotest.(check int) "empty graph has no links" 0
    (List.length (all_links g))

let test_add_device () =
  let g = empty () in
  let d = make_device ~id:"test-rtr" ~role:CoreRouter
    ~site:"dc-east" ~loopback:"10.0.0.1" ~vendor:"cisco" in
  let g = add_device g d in
  Alcotest.(check int) "graph has 1 device" 1
    (List.length (all_devices g));
  match find_device g "test-rtr" with
  | Some found -> Alcotest.(check string) "device id matches" "test-rtr" found.id
  | None -> Alcotest.fail "device not found"

let test_duplicate_device_rejected () =
  let g = empty () in
  let d = make_device ~id:"rtr" ~role:CoreRouter
    ~site:"dc-east" ~loopback:"10.0.0.1" ~vendor:"cisco" in
  let g = add_device g d in
  Alcotest.check_raises "duplicate device raises"
    (Invalid_argument "Device rtr already exists")
    (fun () -> ignore (add_device g d))

let test_add_link () =
  let g = build_linear_topology () in
  Alcotest.(check int) "graph has 3 devices" 3
    (List.length (all_devices g));
  Alcotest.(check int) "graph has 2 links" 2
    (List.length (all_links g))

let test_link_unknown_device () =
  let g = empty () in
  let d = make_device ~id:"A" ~role:CoreRouter
    ~site:"dc-east" ~loopback:"10.0.0.1" ~vendor:"cisco" in
  let g = add_device g d in
  Alcotest.check_raises "link to unknown device raises"
    Not_found
    (fun () -> ignore (add_link g ~src:"A" ~dst:"UNKNOWN"
      (make_link ~id:"A--UNKNOWN" ())))

let test_neighbors () =
  let g = build_linear_topology () in
  let n = neighbors g "B" in
  Alcotest.(check int) "B has 2 neighbors" 2 (List.length n);
  let neighbor_ids = List.map fst n |> List.sort String.compare in
  Alcotest.(check (list string)) "neighbors are A and C"
    ["A"; "C"] neighbor_ids

let test_devices_at_site () =
  let g = build_linear_topology () in
  let dc_east = devices_at_site g "dc-east" in
  Alcotest.(check int) "dc-east has 2 devices" 2
    (List.length dc_east)

(* ============================================================
   Path Computation Tests
   ============================================================ *)

let test_shortest_path_linear () =
  let g = build_linear_topology () in
  match shortest_path g ~src:"A" ~dst:"C"
    ~weight:(fun l -> l.metrics.latency_us) with
  | Some (path, cost) ->
    Alcotest.(check int) "path has 2 hops" 2 (List.length path);
    Alcotest.(check (float 0.01)) "total latency is 150μs" 150.0 cost
  | None ->
    Alcotest.fail "expected a path from A to C"

let test_shortest_path_diamond_picks_fast () =
  let g = build_diamond_topology () in
  match shortest_path g ~src:"A" ~dst:"D"
    ~weight:(fun l -> l.metrics.latency_us) with
  | Some (path, cost) ->
    (* Should pick A -> B -> D (100μs) over A -> C -> D (400μs) *)
    Alcotest.(check (float 0.01)) "picks fast path (100μs)" 100.0 cost;
    let path_nodes = List.map fst path in
    Alcotest.(check (list string)) "path goes through B"
      ["B"; "D"] path_nodes
  | None ->
    Alcotest.fail "expected a path from A to D"

let test_no_path_disconnected () =
  let g = empty () in
  let g = add_device g (make_device ~id:"X" ~role:CoreRouter
    ~site:"dc-east" ~loopback:"10.0.0.1" ~vendor:"cisco") in
  let g = add_device g (make_device ~id:"Y" ~role:CoreRouter
    ~site:"dc-west" ~loopback:"10.0.0.2" ~vendor:"cisco") in
  (* No link between X and Y — they're disconnected *)
  match shortest_path g ~src:"X" ~dst:"Y"
    ~weight:(fun l -> l.metrics.latency_us) with
  | None -> ()  (* Expected: no path *)
  | Some _ -> Alcotest.fail "should not find path in disconnected graph"

let test_path_to_self () =
  let g = build_linear_topology () in
  match shortest_path g ~src:"A" ~dst:"A"
    ~weight:(fun l -> l.metrics.latency_us) with
  | Some (path, cost) ->
    Alcotest.(check int) "path to self has 0 hops" 0 (List.length path);
    Alcotest.(check (float 0.01)) "cost to self is 0" 0.0 cost
  | None ->
    Alcotest.fail "path to self should exist"

(* ============================================================
   Failover Tests — The money tests for a trading network
   ============================================================ *)

let test_failover_after_link_down () =
  let g = build_diamond_topology () in
  (* Before failure: A -> B -> D is the fast path *)
  let before = shortest_path g ~src:"A" ~dst:"D"
    ~weight:(fun l -> l.metrics.latency_us) in
  (match before with
   | Some (_, cost) ->
     Alcotest.(check (float 0.01)) "before: fast path 100μs" 100.0 cost
   | None -> Alcotest.fail "should have path before failure");

  (* Kill the A -- B link *)
  let g = update_link_state g ~src:"A" ~dst:"B" Down in

  (* After failure: should reroute through A -> C -> D *)
  let after = shortest_path g ~src:"A" ~dst:"D"
    ~weight:(fun l -> l.metrics.latency_us) in
  (match after with
   | Some (path, cost) ->
     Alcotest.(check (float 0.01)) "after: slow path 400μs" 400.0 cost;
     let path_nodes = List.map fst path in
     Alcotest.(check (list string)) "reroutes through C"
       ["C"; "D"] path_nodes
   | None -> Alcotest.fail "should reroute after failure")

let test_failover_both_paths_down () =
  let g = build_diamond_topology () in
  let g = update_link_state g ~src:"A" ~dst:"B" Down in
  let g = update_link_state g ~src:"A" ~dst:"C" Down in
  match shortest_path g ~src:"A" ~dst:"D"
    ~weight:(fun l -> l.metrics.latency_us) with
  | None -> ()  (* Expected: no path when both are down *)
  | Some _ -> Alcotest.fail "should have no path with both links down"

let test_degraded_link_penalty () =
  let g = build_diamond_topology () in
  (* Degrade the fast path's first link *)
  let g = update_link_state g ~src:"A" ~dst:"B" Degraded in
  (* Market data weight doubles latency for degraded links:
     Fast path: 50*2 + 50 = 150μs (with degraded penalty)
     Slow path: 200 + 200 = 400μs
     Fast path still wins even degraded *)
  match compute_path g ~src:"A" ~dst:"D"
    ~traffic_class:MarketData ~constraints:[] with
  | Some result ->
    let path_nodes = List.map fst result.path in
    (* Fast path (150μs degraded) still beats slow path (400μs) *)
    Alcotest.(check (list string)) "still uses fast path when degraded"
      ["B"; "D"] path_nodes
  | None -> Alcotest.fail "should find path"

(* ============================================================
   Traffic Class Tests
   ============================================================ *)

let test_market_data_optimizes_latency () =
  let g = build_diamond_topology () in
  match compute_path g ~src:"A" ~dst:"D"
    ~traffic_class:MarketData ~constraints:[] with
  | Some result ->
    Alcotest.(check (float 0.01)) "market data: 100μs path" 100.0
      result.total_latency_us
  | None -> Alcotest.fail "should find path"

let test_bulk_avoids_saturated_links () =
  let g = build_diamond_topology () in
  (* Saturate the fast path *)
  let saturated_metrics = {
    latency_us = 50.0;
    jitter_us = 10.0;
    utilization = 0.95;  (* 95% utilized — almost full *)
    packet_loss = 0.0;
    last_updated = 0.0;
  } in
  let g = update_metrics g ~src:"A" ~dst:"B" saturated_metrics in
  match compute_path g ~src:"A" ~dst:"D"
    ~traffic_class:BulkTransfer ~constraints:[] with
  | Some result ->
    let path_nodes = List.map fst result.path in
    (* Bulk transfer should avoid saturated fast path, use slow path *)
    Alcotest.(check (list string)) "bulk avoids saturated link"
      ["C"; "D"] path_nodes
  | None -> Alcotest.fail "should find path"

(* ============================================================
   Metric Update Tests
   ============================================================ *)

let test_update_metrics () =
  let g = build_linear_topology () in
  let new_metrics = {
    latency_us = 500.0;
    jitter_us = 50.0;
    utilization = 0.8;
    packet_loss = 0.01;
    last_updated = 1000.0;
  } in
  let g = update_metrics g ~src:"A" ~dst:"B" new_metrics in
  match find_link g ~src:"A" ~dst:"B" with
  | Some link ->
    Alcotest.(check (float 0.01)) "latency updated" 500.0
      link.metrics.latency_us;
    Alcotest.(check (float 0.01)) "utilization updated" 0.8
      link.metrics.utilization
  | None -> Alcotest.fail "link should exist"

(* ============================================================
   JSON Serialization Tests
   ============================================================ *)

let test_json_roundtrip () =
  let g = build_linear_topology () in
  let json = to_yojson g in
  match of_yojson json with
  | Ok g2 ->
    Alcotest.(check int) "same device count"
      (List.length (all_devices g)) (List.length (all_devices g2));
    Alcotest.(check int) "same link count"
      (List.length (all_links g)) (List.length (all_links g2))
  | Error msg ->
    Alcotest.fail (Printf.sprintf "JSON roundtrip failed: %s" msg)

(* ============================================================
   Constraint Tests
   ============================================================ *)

let test_max_hops_constraint () =
  let g = build_diamond_topology () in
  (* Only allow 1 hop — no path from A to D can satisfy this *)
  match compute_path g ~src:"A" ~dst:"D"
    ~traffic_class:MarketData ~constraints:[MaxHops 1] with
  | None -> ()  (* Expected: no 1-hop path from A to D *)
  | Some _ -> Alcotest.fail "should not find path with max 1 hop"

let test_max_latency_constraint () =
  let g = build_diamond_topology () in
  (* Max 150μs — only the fast path (100μs) qualifies *)
  match compute_path g ~src:"A" ~dst:"D"
    ~traffic_class:MarketData ~constraints:[MaxLatency 150.0] with
  | Some result ->
    let path_nodes = List.map fst result.path in
    Alcotest.(check (list string)) "picks fast path under latency budget"
      ["B"; "D"] path_nodes
  | None -> Alcotest.fail "fast path should fit within 150μs"

let test_avoid_site_constraint () =
  let g = build_diamond_topology () in
  (* Avoid dc-east — B is in dc-east, so must go through C (dc-west) *)
  match compute_path g ~src:"A" ~dst:"D"
    ~traffic_class:MarketData ~constraints:[AvoidSite "dc-east"] with
  | Some result ->
    let path_nodes = List.map fst result.path in
    Alcotest.(check (list string)) "avoids dc-east, routes through C"
      ["C"; "D"] path_nodes
  | None ->
    (* A is also in dc-east, but AvoidSite checks intermediate nodes
       in the path, not the source. This might fail if A is in the path
       as an intermediate node. Let's accept None as valid here. *)
    ()

(* ============================================================
   Test Runner
   ============================================================ *)

let () =
  Alcotest.run "TradeNet SDN Controller" [
    "graph_construction", [
      Alcotest.test_case "empty graph" `Quick test_empty_graph;
      Alcotest.test_case "add device" `Quick test_add_device;
      Alcotest.test_case "duplicate device rejected" `Quick test_duplicate_device_rejected;
      Alcotest.test_case "add link" `Quick test_add_link;
      Alcotest.test_case "link to unknown device" `Quick test_link_unknown_device;
      Alcotest.test_case "neighbors" `Quick test_neighbors;
      Alcotest.test_case "devices at site" `Quick test_devices_at_site;
    ];
    "path_computation", [
      Alcotest.test_case "shortest path (linear)" `Quick test_shortest_path_linear;
      Alcotest.test_case "shortest path (diamond picks fast)" `Quick test_shortest_path_diamond_picks_fast;
      Alcotest.test_case "no path (disconnected)" `Quick test_no_path_disconnected;
      Alcotest.test_case "path to self" `Quick test_path_to_self;
    ];
    "failover", [
      Alcotest.test_case "reroute after link failure" `Quick test_failover_after_link_down;
      Alcotest.test_case "no path when all links down" `Quick test_failover_both_paths_down;
      Alcotest.test_case "degraded link penalty" `Quick test_degraded_link_penalty;
    ];
    "traffic_classes", [
      Alcotest.test_case "market data optimizes latency" `Quick test_market_data_optimizes_latency;
      Alcotest.test_case "bulk transfer avoids saturated links" `Quick test_bulk_avoids_saturated_links;
    ];
    "metrics", [
      Alcotest.test_case "update link metrics" `Quick test_update_metrics;
    ];
    "serialization", [
      Alcotest.test_case "JSON roundtrip" `Quick test_json_roundtrip;
    ];
    "constraints", [
      Alcotest.test_case "max hops constraint" `Quick test_max_hops_constraint;
      Alcotest.test_case "max latency constraint" `Quick test_max_latency_constraint;
      Alcotest.test_case "avoid site constraint" `Quick test_avoid_site_constraint;
    ];
  ]

File: sdn-controller/tests/test_network_graph.py
"""Unit tests for the network graph and path computation.

Ported from the original OCaml Alcotest suite (test_network_graph.ml). These
verify the core data structures and algorithms the SDN controller relies on:

  1. Graph construction (adding devices and links)
  2. Path computation (Dijkstra correctness)
  3. Failover behavior (path after link failure)
  4. Traffic-class-specific routing (different weights)
  5. Edge cases (no path, single node, disconnected graph)
"""

import pytest

from tradenet_sdn.network_graph import (
    Device,
    DeviceRole,
    Link,
    LinkMetrics,
    LinkState,
    NetworkGraph,
)
from tradenet_sdn.path_engine import (
    AvoidSite,
    MaxHops,
    MaxLatency,
    TrafficClass,
    compute_path,
)

# ============================================================
# Test helpers
# ============================================================

DEFAULT_METRICS = LinkMetrics(
    latency_us=100.0, jitter_us=10.0, utilization=0.1, packet_loss=0.0, last_updated=0.0
)


def make_device(id, role, site, loopback, vendor):
    return Device(id=id, name=id, role=role, site=site, loopback=loopback,
                  vendor=vendor, asn=65001)


def make_link(id, latency=100.0, bw=10000, cost=10):
    return Link(
        id=id,
        src_interface="eth0",
        dst_interface="eth0",
        bandwidth_mbps=bw,
        state=LinkState.UP,
        metrics=LinkMetrics(latency_us=latency, jitter_us=10.0, utilization=0.1,
                            packet_loss=0.0, last_updated=0.0),
        ospf_cost=cost,
    )


def latency_weight(link):
    return link.metrics.latency_us


def build_linear_topology():
    """A simple 3-node linear topology: A -- B -- C."""
    g = NetworkGraph.empty()
    g = g.add_device(make_device("A", DeviceRole.CORE_ROUTER, "dc-east", "10.0.0.1", "cisco"))
    g = g.add_device(make_device("B", DeviceRole.SPINE_SWITCH, "dc-east", "10.0.0.2", "arista"))
    g = g.add_device(make_device("C", DeviceRole.EDGE_ROUTER, "colo-nyse", "10.0.0.3", "juniper"))
    g = g.add_link("A", "B", make_link("A--B", latency=50.0, cost=5))
    g = g.add_link("B", "C", make_link("B--C", latency=100.0, cost=10))
    return g


def build_diamond_topology():
    """A diamond: A --[fast]--> B --> D, A --[slow]--> C --> D.

    Tests that Dijkstra picks the faster path.
    """
    g = NetworkGraph.empty()
    g = g.add_device(make_device("A", DeviceRole.CORE_ROUTER, "dc-east", "10.0.0.1", "cisco"))
    g = g.add_device(make_device("B", DeviceRole.CORE_ROUTER, "dc-east", "10.0.0.2", "cisco"))
    g = g.add_device(make_device("C", DeviceRole.CORE_ROUTER, "dc-west", "10.0.0.3", "cisco"))
    g = g.add_device(make_device("D", DeviceRole.EDGE_ROUTER, "colo-nyse", "10.0.0.4", "juniper"))
    # Fast path: A -> B -> D (50 + 50 = 100μs)
    g = g.add_link("A", "B", make_link("A--B", latency=50.0, cost=5))
    g = g.add_link("B", "D", make_link("B--D", latency=50.0, cost=5))
    # Slow path: A -> C -> D (200 + 200 = 400μs)
    g = g.add_link("A", "C", make_link("A--C", latency=200.0, cost=20))
    g = g.add_link("C", "D", make_link("C--D", latency=200.0, cost=20))
    return g


# ============================================================
# Graph construction
# ============================================================


def test_empty_graph():
    g = NetworkGraph.empty()
    assert len(g.all_devices()) == 0
    assert len(g.all_links()) == 0


def test_add_device():
    g = NetworkGraph.empty()
    d = make_device("test-rtr", DeviceRole.CORE_ROUTER, "dc-east", "10.0.0.1", "cisco")
    g = g.add_device(d)
    assert len(g.all_devices()) == 1
    found = g.find_device("test-rtr")
    assert found is not None
    assert found.id == "test-rtr"


def test_duplicate_device_rejected():
    g = NetworkGraph.empty()
    d = make_device("rtr", DeviceRole.CORE_ROUTER, "dc-east", "10.0.0.1", "cisco")
    g = g.add_device(d)
    with pytest.raises(ValueError, match="Device rtr already exists"):
        g.add_device(d)


def test_add_link():
    g = build_linear_topology()
    assert len(g.all_devices()) == 3
    assert len(g.all_links()) == 2


def test_link_unknown_device():
    g = NetworkGraph.empty()
    d = make_device("A", DeviceRole.CORE_ROUTER, "dc-east", "10.0.0.1", "cisco")
    g = g.add_device(d)
    with pytest.raises(KeyError):
        g.add_link("A", "UNKNOWN", make_link("A--UNKNOWN"))


def test_neighbors():
    g = build_linear_topology()
    n = g.neighbors("B")
    assert len(n) == 2
    neighbor_ids = sorted(dst for dst, _link in n)
    assert neighbor_ids == ["A", "C"]


def test_devices_at_site():
    g = build_linear_topology()
    dc_east = g.devices_at_site("dc-east")
    assert len(dc_east) == 2


# ============================================================
# Path computation
# ============================================================


def test_shortest_path_linear():
    g = build_linear_topology()
    found = g.shortest_path("A", "C", latency_weight)
    assert found is not None
    path, cost = found
    assert len(path) == 2  # 2 hops
    assert cost == pytest.approx(150.0, abs=0.01)  # 50 + 100


def test_shortest_path_diamond_picks_fast():
    g = build_diamond_topology()
    found = g.shortest_path("A", "D", latency_weight)
    assert found is not None
    path, cost = found
    # Should pick A -> B -> D (100μs) over A -> C -> D (400μs)
    assert cost == pytest.approx(100.0, abs=0.01)
    assert [node for node, _link in path] == ["B", "D"]


def test_no_path_disconnected():
    g = NetworkGraph.empty()
    g = g.add_device(make_device("X", DeviceRole.CORE_ROUTER, "dc-east", "10.0.0.1", "cisco"))
    g = g.add_device(make_device("Y", DeviceRole.CORE_ROUTER, "dc-west", "10.0.0.2", "cisco"))
    # No link between X and Y — they're disconnected
    assert g.shortest_path("X", "Y", latency_weight) is None


def test_path_to_self():
    g = build_linear_topology()
    found = g.shortest_path("A", "A", latency_weight)
    assert found is not None
    path, cost = found
    assert len(path) == 0
    assert cost == pytest.approx(0.0, abs=0.01)


# ============================================================
# Failover — the money tests for a trading network
# ============================================================


def test_failover_after_link_down():
    g = build_diamond_topology()
    # Before failure: A -> B -> D is the fast path (100μs)
    before = g.shortest_path("A", "D", latency_weight)
    assert before is not None
    assert before[1] == pytest.approx(100.0, abs=0.01)

    # Kill the A -- B link
    g2 = g.update_link_state("A", "B", LinkState.DOWN)

    # After failure: should reroute through A -> C -> D (400μs)
    after = g2.shortest_path("A", "D", latency_weight)
    assert after is not None
    path, cost = after
    assert cost == pytest.approx(400.0, abs=0.01)
    assert [node for node, _link in path] == ["C", "D"]


def test_failover_both_paths_down():
    g = build_diamond_topology()
    g = g.update_link_state("A", "B", LinkState.DOWN)
    g = g.update_link_state("A", "C", LinkState.DOWN)
    assert g.shortest_path("A", "D", latency_weight) is None


def test_degraded_link_penalty():
    g = build_diamond_topology()
    # Degrade the fast path's first link. Market-data weight doubles latency for
    # degraded links: fast path = 50*2 + 50 = 150μs, slow path = 400μs.
    # Fast path still wins even degraded.
    g = g.update_link_state("A", "B", LinkState.DEGRADED)
    result = compute_path(g, "A", "D", TrafficClass.MARKET_DATA, constraints=[])
    assert result is not None
    assert [node for node, _link in result.path] == ["B", "D"]


# ============================================================
# Traffic classes
# ============================================================


def test_market_data_optimizes_latency():
    g = build_diamond_topology()
    result = compute_path(g, "A", "D", TrafficClass.MARKET_DATA, constraints=[])
    assert result is not None
    assert result.total_latency_us == pytest.approx(100.0, abs=0.01)


def test_bulk_avoids_saturated_links():
    g = build_diamond_topology()
    # Saturate the fast path (95% utilized)
    saturated = LinkMetrics(latency_us=50.0, jitter_us=10.0, utilization=0.95,
                            packet_loss=0.0, last_updated=0.0)
    g = g.update_metrics("A", "B", saturated)
    result = compute_path(g, "A", "D", TrafficClass.BULK_TRANSFER, constraints=[])
    assert result is not None
    # Bulk transfer should avoid the saturated fast path, use slow path
    assert [node for node, _link in result.path] == ["C", "D"]


# ============================================================
# Metric updates
# ============================================================


def test_update_metrics():
    g = build_linear_topology()
    new_metrics = LinkMetrics(latency_us=500.0, jitter_us=50.0, utilization=0.8,
                              packet_loss=0.01, last_updated=1000.0)
    g = g.update_metrics("A", "B", new_metrics)
    link = g.find_link("A", "B")
    assert link is not None
    assert link.metrics.latency_us == pytest.approx(500.0, abs=0.01)
    assert link.metrics.utilization == pytest.approx(0.8, abs=0.01)


# ============================================================
# JSON serialization
# ============================================================


def test_json_roundtrip():
    g = build_linear_topology()
    data = g.to_dict()
    g2 = NetworkGraph.from_dict(data)
    assert len(g.all_devices()) == len(g2.all_devices())
    assert len(g.all_links()) == len(g2.all_links())


# ============================================================
# Constraints
# ============================================================


def test_max_hops_constraint():
    g = build_diamond_topology()
    # Only allow 1 hop — no path from A to D can satisfy this
    result = compute_path(g, "A", "D", TrafficClass.MARKET_DATA, constraints=[MaxHops(1)])
    assert result is None


def test_max_latency_constraint():
    g = build_diamond_topology()
    # Max 150μs — only the fast path (100μs) qualifies
    result = compute_path(g, "A", "D", TrafficClass.MARKET_DATA,
                          constraints=[MaxLatency(150.0)])
    assert result is not None
    assert [node for node, _link in result.path] == ["B", "D"]


def test_avoid_site_constraint():
    g = build_diamond_topology()
    # Avoid dc-east — B is in dc-east, so must route through C (dc-west).
    result = compute_path(g, "A", "D", TrafficClass.MARKET_DATA,
                          constraints=[AvoidSite("dc-east")])
    if result is not None:
        assert [node for node, _link in result.path] == ["C", "D"]
    # If None: A is also in dc-east; AvoidSite checks intermediate path nodes.
    # Accepting None as valid here, matching the original test.
